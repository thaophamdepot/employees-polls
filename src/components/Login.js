import * as React from "react";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import CssBaseline from "@mui/material/CssBaseline";
import TextField from "@mui/material/TextField";
import Visibility from "@mui/icons-material/Visibility";
import VisibilityOff from "@mui/icons-material/VisibilityOff";
import Typography from "@mui/material/Typography";
import Container from "@mui/material/Container";
import { IconButton, InputAdornment, MenuItem, Select } from "@mui/material";
import { createTheme, ThemeProvider } from "@mui/material/styles";
import { connect } from "react-redux";
import { useNavigate, useLocation, useParams } from "react-router-dom";

import { setAuthedUser } from "../actions/authedUser";

const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};

const theme = createTheme();

const Login = (props) => {
  const navigate = useNavigate();
  const search = useLocation().search;

  const [usernameValue, setUsernameValue] = React.useState("");
  const [passwordValue, setPasswordValue] = React.useState("");
  const [showPassword, setShowPassword] = React.useState(false);
  const [error, setError] = React.useState(false);

  const handleClickShowPassword = () => setShowPassword(!showPassword);
  const { users } = props;

  const handleUsernameChange = (e) => {
    setUsernameValue(e.target.value);
  };

  const handlePasswordChange = (e) => {
    setPasswordValue(e.target.value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    const username = usernameValue;
    const password = passwordValue;

    if (username === "" || users[username].password !== password) {
      setError(true);
    } else if (users[username]) {
      new Promise((res, rej) => {
        setTimeout(() => res(), 300);
      }).then(() => {
        props.dispatch(setAuthedUser(username));
        setError(false);
        const redirectTo = new URLSearchParams(search).get("redirectTo");
        redirectTo && redirectTo !== ""
          ? navigate(redirectTo)
          : navigate("/");
      });
    }
  };

  return (
    <ThemeProvider theme={theme}>
      <Container component="main" maxWidth="xs">
        <CssBaseline />
        <Box
          sx={{
            marginTop: 8,
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
          }}
        >
          <Typography component="h1" variant="h5" fontWeight={"bold"}>
            Log In
          </Typography>
          {error && (
            <Typography
              color="error"
              className={"Error"}
              data-testid="error-header"
            >
              &nbsp;{"User name and password is incorrect!"}&nbsp;
            </Typography>
          )}
          <Box component="form" onSubmit={handleSubmit} sx={{ mt: 1 }}>
            <label>User:</label>
            <Select
              labelId="demo-simple-select-required-label"
              id="demo-simple-select-required"
              label="Username"
              value={usernameValue}
              onChange={handleUsernameChange}
            >
              {Object.keys(users).map((item) => (
                <MenuItem key={users[item].id} value={users[item].id}>
                  {users[item].name}
                </MenuItem>
              ))}
            </Select>
            <TextField
              margin="normal"
              fullWidth
              name="password"
              label="Password"
              value={passwordValue}
              onChange={handlePasswordChange}
              type={showPassword ? "text" : "password"}
              id="password"
              data-testid="password"
              autoComplete="current-password"
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <IconButton
                      aria-label="toggle password visibility"
                      onClick={handleClickShowPassword}
                    >
                      {showPassword ? <Visibility /> : <VisibilityOff />}
                    </IconButton>
                  </InputAdornment>
                ),
              }}
            />
            <Button
              type="submit"
              fullWidth
              variant="contained"
              sx={{ mt: 3, mb: 2 }}
              data-testid="sign-in"
              color="primary"
            >
              <span style={{fontWeight: "bold"}}>Log In</span>
            </Button>
          </Box>
        </Box>
      </Container>
    </ThemeProvider>
  );
};

function mapStateToProps({ authedUser, users }) {
  return {
    authedUser: authedUser !== null,
    users,
  };
}

export default withRouter(connect(mapStateToProps)(Login));
