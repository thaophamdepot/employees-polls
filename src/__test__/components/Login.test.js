import { render, screen, fireEvent } from "@testing-library/react";
import Login from "../../components/Login";
import configureStore from "redux-mock-store";
import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { Router } from "react-router";
import { createMemoryHistory } from "history";

const usersData = {
  "test-user-id": {
    id: "test-user-id",
    password: "test-pass",
    name: "Test User",
    avatarURL: "",
    answers: {
      "test-question-id-one": "optionOne",
    },
    questions: ["test-question-id-two"],
  },
  "test-user-id-one": {
    id: "test-user-id-one",
    password: "test-pass",
    name: "Test User",
    avatarURL: "",
    answers: {
      "test-question-id-two": "optionOne",
    },
    questions: ["test-question-id-three"],
  },
};

const mockStore = configureStore([]);

describe("Check username, password and submit button is present.", () => {
  let store;
  let component;

  let users = usersData;
  const history = createMemoryHistory();
  beforeEach(() => {
    store = mockStore({
      authedUser: "test-user-id",
      users: users,
    });

    component = renderer.create(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
  });

  test("Check username field is there in the Login component.", () => {
    component = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
    var username = component.getAllByText(/Username/);
    expect(username.length).toEqual(1);
  });

  test("Check password field is there in the Login component.", () => {
    component = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
    var password = component.getByTestId("password");
    expect(password).toBeInTheDocument();
  });

  test("Check Log In button is there in the Login component.", () => {
    component = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
    var LoginButton = component.getByTestId("sign-in");
    expect(LoginButton).toBeInTheDocument();
  });
});

describe("Login component - Snapshot testing", () => {
  let store;
  let component;

  let users = usersData;
  const history = createMemoryHistory();
  beforeEach(() => {
    store = mockStore({
      authedUser: "test-user-id",
      users: users,
    });

    component = renderer.create(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
  });

  test("renders Login component with screen.debug() in UI", async () => {
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
    screen.debug();
  });

  test("renders Login component with toMatchSnapshot in UI", () => {
    expect(component).toMatchSnapshot();
  });
});

describe("FireEvent testing with Login component", () => {
  let store;
  let component;

  let users = usersData;
  const history = createMemoryHistory();
  beforeEach(() => {
    store = mockStore({
      authedUser: "test-user-id",
      users: users,
    });

    component = renderer.create(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
  });

  test("If we do not pick one user from dropdown list, then it should popup the error message", () => {
    component = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Login />
        </Router>
      </Provider>
    );
    var password = component.getByTestId("password").querySelector("input");
    fireEvent.change(password, { target: { value: "123456" } });
    var LoginButton = component.getByTestId("sign-in");
    fireEvent.click(LoginButton);
    expect(component.getByTestId("error-header")).toBeInTheDocument();
  });
});
