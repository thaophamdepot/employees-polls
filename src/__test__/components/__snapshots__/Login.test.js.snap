// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component - Snapshot testing renders Login component with toMatchSnapshot in UI 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
>
  <div
    className="MuiBox-root css-binzgt"
  >
    <h1
      className="MuiTypography-root MuiTypography-h5 css-bezmlo-MuiTypography-root"
    >
      Log In
    </h1>
    <form
      className="MuiBox-root css-164r41r"
      onSubmit={[Function]}
    >
      <label>
        User:
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-controls=":r8:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="demo-simple-select-required-label demo-simple-select-required"
          className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="demo-simple-select-required"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="combobox"
          tabIndex={0}
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </div>
        <input
          aria-hidden={true}
          aria-invalid={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Username
            </span>
          </legend>
        </fieldset>
      </div>
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        data-testid="password"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="current-password"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
            value=""
          />
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="toggle password visibility"
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="VisibilityOffIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
        data-testid="sign-in"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Log In
        </span>
      </button>
    </form>
  </div>
</main>
`;
